src/global/utils/RandomUtil.java

```diff
@@ -0,0 +1,113 @@
+    }
+
+    public static int[] getSplitDecimal(double num) {
+        String[] lowSplit = String.valueOf(num).split("\\.");
+        return new int[]{Integer.parseInt(lowSplit[0]), Integer.parseInt(lowSplit[1])};
+    }
+
+    public static boolean followsUniformDistribution(ArrayList<Integer> values, int lower, int upper) {
+        final int NUM_BINS = getNumBins(lower, upper);
+
+        long[] observedCounts = new long[NUM_BINS];
+        for (int value : values) {
+            int binNum = assignedBinNum(value, lower, upper, NUM_BINS);
+            if (binNum == NO_BIN) return false;
+            observedCounts[assignedBinNum(value, lower, upper, NUM_BINS)]++;
+        }
+
+        double[] expected = new double[NUM_BINS];
+        int expectedFreq = values.size() / NUM_BINS;
+        Arrays.fill(expected, expectedFreq);
+
+        ChiSquareTest cst = new ChiSquareTest();
+        // function returns false if we fail to reject null hypothesis, which implies the values are uniformly distributed
+        return !cst.chiSquareTest(expected, observedCounts, ALPHA_LEVEL);
+    }
+
+    public static boolean followsUniformDistribution(ArrayList<Double> values, double lower, double upper) {
+        final int NUM_BINS = getNumBins(lower, upper);
+
+        long[] observedCounts = new long[NUM_BINS];
+        for (double value : values) {
+            int binNum = assignedBinNum(value, lower, upper, NUM_BINS);
+            if (binNum == NO_BIN) return false;
+            observedCounts[assignedBinNum(value, lower, upper, NUM_BINS)]++;
+        }
+
+        double[] expected = new double[NUM_BINS];
+        int expectedFreq = values.size() / NUM_BINS;
+        Arrays.fill(expected, expectedFreq);
+
+        ChiSquareTest cst = new ChiSquareTest();
+        // function returns false if we fail to reject null hypothesis, which implies the values are uniformly distributed
+        return !cst.chiSquareTest(expected, observedCounts, ALPHA_LEVEL);
+    }
+
+    public static int assignedBinNum(double value, int lower, int upper, int numBins) {
+        int range = upper - lower;
+        return assignedBinNumHelper(range, value, lower, upper, numBins);
+    }
+
+    public static int assignedBinNum(double value, double lower, double upper, int numBins) {
+        double range = upper - lower;
+        return assignedBinNumHelper(range, value, lower, upper, numBins);
+    }
+
+    public static int assignedBinNumHelper(double range, double value, double lower, double upper, int numBins) {
+        double gap = range / numBins;
+        double binLower = lower;
+        int count = 0;
+
+        assertWithinRange(value, lower, upper, "One or more of your randomly generated numbers fall outside of the required range.");
+
+        while (binLower < upper) {
+            double binUpper = binLower + gap;
+            if (binLower <= value && value < binUpper)
+                return count;
+            count++;
+            binLower = binUpper;
+        }
```

Comment:
This works but for efficiency's sake, I think since this will traverse every single bin one at a time from the start it might hamper the runtime sometimes. I think if we leave the bins as a constant then technically it is O(1) but in case we change it, I want to see if there is a "mathy" way to make this quicker.

Reply:
Made #11 to resolve this 👍🏾